engine/consensus/sealing/engine.go

```diff
@@ -36,96 +34,68 @@ type (
 	EventSink chan *Event // Channel to push pending events
 )
 
-// Engine is a wrapper for sealing `Core` which implements logic for
+// Engine is a wrapper for approval processing `Core` which implements logic for
 // queuing and filtering network messages which later will be processed by sealing engine.
 // Purpose of this struct is to provide an efficient way how to consume messages from network layer and pass
 // them to `Core`. Engine runs 2 separate gorourtines that perform pre-processing and consuming messages by Core.
 type Engine struct {
 	unit                                 *engine.Unit
+	core                                 sealing.SealingCore
+	workerPool                           *workerpool.WorkerPool
```

Comment:
I'd like to revisit [Leo's comment](https://github.com/onflow/flow-go/pull/736#discussion_r637303532) into broader context of the `WorkerPool`:
* As you can see from Leo's question, the worker pool adds some complexity to the overall processing logic
* I feel that the worker pool is (unnecessarily) replicating some logic which we already natively have in the engine: 
    * it has a (single!) queue for inbound work
    * a notification system to wake up routines 
   
  We already have these, though, more specialized for the specific engine use-case and more composable. 

Lets take the example of an inbound approval:
1. The networking layer will call into the `MessageHander` and put the approval into the  `pendingApprovals` queue 
2. the Engine's `loop` routine will pick up the proposal and `Submit` the task for processing the approval to the worker pool (put it into the pool's internally a channel `taskQueue`)
3. The worker pool has an internal `task dispatcher` routine, which will pick up the task (from `taskQueue` channel) and potentially add the task to the pools internal queue of waiting tasks (`waitingQueue`)
4. When a worker becomes available, the `task dispatcher` will push the task to the worker
5. worker starts task 

In my opinion, steps 2, 3 and 4 are not needed. I think what we need is some sort of queue for incoming incorporated results. Atm, we are using the worker pool's internal task queue for it which creates all these extra layers for the inbound information to flow through. Furthermore, in my opinion, it additionally obfuscates the message flow. 

Specifically, the ability to prioritize messages is undermined by the worker pool; specifically by the fact that the worker pool stores _all_ waiting tasks into a single queue. For this reason, the priority of a message is determined by when the task for processing the information was created. Once the task is in the worker pool's queue, higher-priority information that arrived later cannot be given priority anymore. I don't think this is a problem for the sealing engine in particular. Though, I think it is an inflexibility of the `WorkerPool`-based design in general, which I think would be great to avoid. 

### Suggestion 

I am fairly optimistic that it is straightforward to remove the `WorkerPool` entirely. As argued above, I think we have all the building blocks we need already. The central trick in my suggestion is: `n` basic go-routines routines are executing the engine's `loop()` method concurrently. 

Details:
1. Include a dedicated inbound queue which holds `IncorporatedResult`s
    ```golang
    pendingIncorporatedResults *fifoqueue.FifoQueue
    ```
2. Remove the `WorkerPool` entirely     
3. In method https://github.com/onflow/flow-go/blob/a1208aa8ea49ac8a3115cf7840c7ec9632a9a63a/engine/consensus/sealing/engine.go#L317-L320 I would
    - change the `e.workerPool.Submit` to 
      ```golang
      go func(){ ..
      ```
      I think this is ok because the rate at which `OnBlockIncorporated` is called is fundamentally limited by consensus's block production rate. Furthermore, this callback is invoked by HotStuff, _not_ with blocks from the networking layer. All blocks must first pass through the compliance engine and HotStuff, which also have a limited processing rate. Therefore, the risk of creating an enormous backlog of go-routines here is quite impossible. Lastly, the order in which we process incorporated blocks is largely irrelevant (except for some mild performance impact if the blocks get too much out of order -- however, blocks getting out of order should be super rate due to the very limited block rate). 
    - for all the results in the payload https://github.com/onflow/flow-go/blob/a1208aa8ea49ac8a3115cf7840c7ec9632a9a63a/engine/consensus/sealing/engine.go#L339-L341 
      we create the `IncorporatedResult` and add queue it in `pendingIncorporatedResults` (and trigger the Notifier at the end)
4. In https://github.com/onflow/flow-go/blob/a1208aa8ea49ac8a3115cf7840c7ec9632a9a63a/engine/consensus/sealing/engine.go#L228 you can directly call into `Core`'s business logic functions, because `loop` is executed by worker routines. 
5. Same for `OnFinalizedBlock`: create a dedicated queue to hold the `finalizedBlockID` values and process them as part of `loop`

Reply:
I have introduced a few changes as you suggested and merged message store approach for approvals + added extra queues for finalized/incorporated results